WooCommerce/Classes/POS/ViewModels/TotalsViewModel.swift

```diff
@@ -341,25 +347,37 @@ private extension TotalsViewModel {
     /// - Returns: PointOfSaleCardPresentPaymentMessageType
     func mapCardPresentPaymentEventToMessageType(_ event: CardPresentPaymentEvent) -> PointOfSaleCardPresentPaymentMessageType? {
         guard case let .show(eventDetails) = event,
-              case let .message(messageType) = eventDetails.pointOfSalePresentationStyle else {
+              case let .message(messageType) = presentationStyle(for: eventDetails) else {
             return nil
         }
 
-        switch messageType {
-        case .paymentSuccess(var viewModel):
-            guard let formattedOrderTotalPrice else {
-                return messageType
-            }
-            viewModel.updateMessage(formattedOrderTotal: formattedOrderTotalPrice)
-            return .paymentSuccess(viewModel: viewModel)
-        default:
-            return messageType
+        return messageType
+    }
+
+    func presentationStyle(for eventDetails: CardPresentPaymentEventDetails) -> PointOfSaleCardPresentPaymentEventPresentationStyle? {
+        PointOfSaleCardPresentPaymentEventPresentationStyleDeterminer.presentationStyle(
+            for: eventDetails,
+            dependencies: presentationStyleDeterminerDependencies)
+    }
+
+    var presentationStyleDeterminerDependencies: PointOfSaleCardPresentPaymentEventPresentationStyleDeterminer.Dependencies {
+        PointOfSaleCardPresentPaymentEventPresentationStyleDeterminer.Dependencies(
+            tryPaymentAgainBackToCheckoutAction: cancelThenCollectPayment,
+            nonRetryableErrorExitAction: cancelThenCollectPayment,
+            formattedOrderTotalPrice: formattedOrderTotalPrice)
+    }
+
+    func cancelThenCollectPayment() {
```

Comment:
So, the two buttons are subtly different. The retry button retries the payment as late in the flow as it can. Depending on the error, it might not require another card tap to retry. That's why it doesn't say `Try another payment method`.

Note that in `PointOfSaleCardPresentPaymentEventPresentationStyleDeterminer.presentationStyle(for:dependencies:)` we use `retryAction` from the payment event for the try payment again button:

```
                    return .message(.paymentError(
                        viewModel: PointOfSaleCardPresentPaymentErrorMessageViewModel(
                            error: error,
                            tryPaymentAgainButtonAction: retryAction,
                            backToCheckoutButtonAction: dependencies.tryPaymentAgainBackToCheckoutAction)))
```

Because the merchant may be nervous about this kind of retry, and just wants to go back, we give them the other button rather than force them into the second attempt.

It might seem confusing to use `cancelThenCollectPayment` to go back; we do it because it guarantees that the card will need to be tapped again. The checkout is strange if you don't prepare the reader for payment, because it just shows the totals, with no action to take.

Reply:
> Note that in PointOfSaleCardPresentPaymentEventPresentationStyleDeterminer.presentationStyle(for:dependencies:) we use retryAction from the payment event for the try payment again button:

Oh, got it. Thanks for the explanation!